src/main/webapp/WEB-INF/view/activity.jsp

```diff
@@ -50,11 +54,20 @@ List<Conversation> conversations = (List<Conversation>) request.getAttribute("co
 <div id="feed">
   <ul>
     <%
-      for (Conversation conversation : conversations) {
-        String title = conversation.getTitle();
+      /* Get the titles of the conversations. */
+       for (Conversation conversation : conversations) {
+         String title = conversation.getTitle();
     %>
-    <li><strong><%= title %></strong></li>
-    <%
+       <li><strong><%= title %></strong></li>
+       <%
+            /* Get the messages of the conversations. */
+         List<Message> messages = messageStore.getMessagesInConversation(conversation.getId());
```

Comment:
nit:
While this works. But what I have seen in the existing code is that the Servlet would set an attribute 'messages' using the messageStore, and the JSP would iterate through the messages. JSP does not directly interact with the messageStore.

But here you are passing the messageStore directly to the JSP. Unclear to me if it would cause some problem. So I am not going to object to what you are doing here.